src/udev/udev-builtin-hwdb.c

```diff
@@ -99,6 +99,17 @@ static int udev_builtin_hwdb_search(sd_device *dev, sd_device *srcdev,
                         last = true;
                 }
 
+                /* The fw node device has character device while the fw unit device doesn't. */
```

Comment:
can't we move this to the udev rules? i.e. simply not issue hwdb lookups on the node devices?

Reply:
Hi @poettering ,

> can't we move this to the udev rules? i.e. simply not issue hwdb lookups on the node devices?

Hm. If we resolved the issue in the domain of udev rule, it would be preferrable.

However, current implementation of hwdb builtin in udev searches recursively till matching any database entry. It's inconvenient in the case that hwdb has no entry corresponding to fw __unit__ device since it's queried in line 14 of `rules.d/50-udev-default.rules.in`:

```
14: ENV{MODALIAS}!="", IMPORT{builtin}="hwdb --subsystem=$env{SUBSYSTEM}"
```

In the case, the search attempts:

1. fw unit device (has modalias, assuming no hwdb entry, continue)
2. fw node device (doesn't have modalias, continue)
3. pci device (has modalias and hwdb entry, exit)

In my opinion, any control flow to cancel recursive search is required for the case. My proposal is hard-code just for IEEE 1394 case. But it would be useful as well to add an option to the builtin to decide the cancel timing, like '--recursive-count=1' for query to fw unit, '--recursive-count=0' for query to fw node, so that usb device can use the option.

Reply:
another idea might be: define a new property that we check when walking up the tree. whenever it is set we abort the walking up the tree.

i.e. ID_HWDB_BARRIER=1 or so. we'd set it for the fw node from a udev rule, and hwdb would look for it and exit the walk. That way the hw subsys specific test remains in udev rules.

Reply:
The idea looks good for the case of fw node device however it doesn't for the case of fw unit, at least in current udev rule.

If applying the idea to fw unit device as well, it might be achieved to cancel walking up to fw node, however it seems to be overplus restriction.

It would be the best to find the way to suppress walking up to pci tree regardless of querying fw node or fw unit...